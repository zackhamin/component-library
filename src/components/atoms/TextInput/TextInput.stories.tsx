import { Story } from '@storybook/react';
import { TextInput, TextInputCombinedProps } from './TextInput';

export default {
	title: 'atoms/TextInput',
	component: TextInput,
};

const Template: Story<TextInputCombinedProps> = (args) => (
	<TextInput {...args} />
);

const defaultProps = {
	id: 'inputID',
	name: 'inputName',
	placeholder: 'Placeholder text',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Disabled = Template.bind({});
Disabled.args = {
	...defaultProps,
	isDisabled: true,
	placeholder: '',
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
