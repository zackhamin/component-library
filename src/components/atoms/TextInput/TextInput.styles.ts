import { hexToRGB } from 'Utils/utils';
import styled from 'styled-components';
import { colors, typography } from 'Styles';

const StyledTextInput = styled.input`
	height: 4.8rem;
	width: 100%;
	padding: 1.6rem;
	font-size: ${typography.size.s3}rem;
	font-family: ${typography.type.body};
	color: ${colors.offBlack};
	border: 1px solid ${colors.offBlack};
	border-radius: 0.4rem;

	&::placeholder {
		color: ${colors.grey500};
	}

	&:disabled {
		background-color: ${colors.lightGrey};
		border-color: ${colors.grey500};
	}

	&:focus-visible {
		outline: none;
		border-color: ${colors.emerald};
		box-shadow: 0px 0px 6px ${hexToRGB(colors.emerald, 0.4)};
	}

	&.is-invalid {
		border-color: ${colors.error};
		// Compensation for icon
		padding-right: 5rem;
	}

	&.is-success {
		// Compensation for icon
		padding-right: 5rem;
	}
`;

const StyledTextInputWrapper = styled.div`
	position: relative;
`;

const StyledTextInputIcon = styled.div`
	position: absolute;
	top: 50%;
	right: 1.5rem;
	transform: translateY(-50%);
`;

export { StyledTextInput, StyledTextInputWrapper, StyledTextInputIcon };
