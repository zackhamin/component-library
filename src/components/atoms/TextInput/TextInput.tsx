import { colors } from 'Styles';
import React from 'react';
import { useInputContext } from 'Hooks/useInputContext';
import { BaseInputProps, ConditionalFormProps } from '@Types/form';
import { Check, Error } from 'Atoms/Icons';
import {
	StyledTextInput,
	StyledTextInputIcon,
	StyledTextInputWrapper,
} from './TextInput.styles';

export interface TextInputProps extends BaseInputProps<HTMLInputElement> {
	type?: 'text' | 'email' | 'tel';
	placeholder?: string;
	value?: string;
	onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export type TextInputCombinedProps = TextInputProps & ConditionalFormProps;

const TextInput = React.forwardRef<HTMLInputElement, TextInputCombinedProps>(
	(
		{
			name,
			id,
			isDisabled,
			isInvalid,
			isRequired,
			isSuccess,
			type = 'text',
			placeholder,
			className,
			value,
			onChange,
			...rest
		},
		ref
	) => {
		const { inputClasses, ctx } = useInputContext({
			id,
			name,
			isInvalid,
			isRequired,
			isSuccess,
			className,
		});

		return (
			<StyledTextInputWrapper>
				<StyledTextInput
					className={inputClasses}
					name={ctx.name}
					id={ctx.id}
					ref={ref}
					type={type}
					placeholder={placeholder}
					aria-invalid={ctx.isInvalid}
					aria-describedby={ctx.isInvalid ? `${ctx.id}--feedback` : undefined}
					required={ctx.isRequired}
					disabled={isDisabled}
					value={value}
					onChange={onChange}
					{...rest}
				/>

				{ctx.isInvalid && !isDisabled && (
					<StyledTextInputIcon>
						<Error fontSize="1.8rem" color={colors.error} data-testid="ERROR" />
					</StyledTextInputIcon>
				)}

				{ctx.isSuccess && !isDisabled && (
					<StyledTextInputIcon>
						<Check
							fontSize="1.8rem"
							color={colors.emerald75}
							data-testid="CHECK"
						/>
					</StyledTextInputIcon>
				)}
			</StyledTextInputWrapper>
		);
	}
);

export { TextInput };
