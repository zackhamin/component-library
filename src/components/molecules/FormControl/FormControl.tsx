import { ConditionalFormProps } from '@Types/form';
import { createContext } from 'react';

interface FormControlContext {
	name?: string;
	id: string;
	isRequired?: boolean;
	isInvalid?: boolean;
	isSuccess?: boolean;
}

interface BaseFormControlProps extends FormControlContext {
	children: React.ReactNode;
	className?: string;
}

export type FormControlProps = BaseFormControlProps & ConditionalFormProps;

export const FormControlContext = createContext<FormControlContext | undefined>(
	undefined
);

const FormControl = ({
	children,
	name,
	id,
	isInvalid,
	isRequired,
	isSuccess,
	className,
}: FormControlProps) => {
	return (
		<FormControlContext.Provider
			value={{ name, id, isRequired, isInvalid, isSuccess }}
		>
			<div className={className}>{children}</div>
		</FormControlContext.Provider>
	);
};

export { FormControl };
