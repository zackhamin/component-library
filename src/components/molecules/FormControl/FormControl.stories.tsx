import { FormError } from 'Atoms/FormError';
import { FormLabel } from 'Atoms/FormLabel';
import { Story } from '@storybook/react';
import { TextInput } from 'Atoms/TextInput';
import { FormControl, FormControlProps } from './FormControl';

export default {
	title: 'molecules/FormControl',
	component: FormControl,
	argTypes: {
		isSuccess: {
			control: {
				type: 'boolean',
			},
		},
		isInvalid: {
			control: {
				type: 'boolean',
			},
		},
	},
};

const Template: Story<FormControlProps> = (args) => (
	<FormControl {...args}>
		<FormLabel>Label</FormLabel>
		<TextInput />
		<FormError>Error Message</FormError>
	</FormControl>
);
const defaultProps = {
	name: 'testName',
	id: 'testId',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Required = Template.bind({});
Required.args = {
	...defaultProps,
	isRequired: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
