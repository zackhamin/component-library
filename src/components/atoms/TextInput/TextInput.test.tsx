import { FormControl } from 'Molecules/FormControl';
import { TextInput } from '../TextInput';

import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps = {
	id: 'inputID',
	name: 'inputName',
	placeholder: 'Placeholder text',
	className: 'additional-class',
};

describe('TextInput', () => {
	it('renders correctly', () => {
		const { container } = render(<TextInput {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});

	it('adds ID, class, type & name to element', () => {
		render(<TextInput {...defaultProps} type="email" />);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('id', 'inputID');
		expect(input).toHaveAttribute('name', 'inputName');
		expect(input).toHaveAttribute('type', 'email');
		expect(input).toHaveClass('additional-class');
		expect(input).not.toHaveAttribute('aria-invalid');
		expect(input).not.toHaveAttribute('aria-describedby');
	});
});

describe('Text input has disabled state', () => {
	it('adds disabled attribute to element', () => {
		render(<TextInput {...defaultProps} isDisabled />);
		expect(screen.getByRole('textbox')).toHaveAttribute('disabled');
	});
});

describe('Text input has invalid state', () => {
	it('renders with invalid class', () => {
		render(<TextInput {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveClass('is-invalid');
	});

	it('adds aria-invalid attribute to element', () => {
		render(<TextInput {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveAttribute('aria-invalid', 'true');
	});

	it('should have aria-describedby attribute based on id', () => {
		render(<TextInput {...defaultProps} isInvalid />);
		expect(screen.getByRole('textbox')).toHaveAttribute(
			'aria-describedby',
			'inputID--feedback'
		);
	});

	it('should display an SVG error icon', () => {
		render(<TextInput {...defaultProps} isInvalid />);
		expect(screen.getAllByTestId('ERROR')).toHaveLength(1);
	});
});

describe('Text input has success state', () => {
	it('renders with success class', () => {
		render(<TextInput {...defaultProps} isSuccess />);
		expect(screen.getByRole('textbox')).toHaveClass('is-success');
	});

	it('does not have aria-invalid & aria-describedby attributes', () => {
		render(<TextInput {...defaultProps} isSuccess />);
		expect(screen.getByRole('textbox')).not.toHaveAttribute('aria-invalid');
		expect(screen.getByRole('textbox')).not.toHaveAttribute('aria-describedby');
	});

	it('should display an SVG error icon', () => {
		render(<TextInput {...defaultProps} isSuccess />);
		expect(screen.getAllByTestId('CHECK')).toHaveLength(1);
	});
});

describe('Text input change handler', () => {
	it('should call onChange on user input', () => {
		const onChangeHandler = jest.fn();
		const inputValue = 'input value';

		render(<TextInput {...defaultProps} onChange={onChangeHandler} />);

		const input = screen.getByRole('textbox');

		fireEvent.change(input, { target: { value: inputValue } });

		expect(onChangeHandler).toHaveBeenCalled();
	});
});

describe('TextArea wrapped in FormControl', () => {
	it('should have the correct ID passed from Form Control', () => {
		render(
			<FormControl id="control">
				<TextInput />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('id', 'control');
	});

	it('should have required attr when isRequired is passed from Form Control', () => {
		render(
			<FormControl id="control" isRequired>
				<TextInput />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('required');
	});

	it('should have invalid attr and class when isInvalid passed from Form Control', () => {
		render(
			<FormControl id="control" isInvalid>
				<TextInput />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveAttribute('aria-invalid', 'true');
		expect(input).toHaveClass('is-invalid');
	});

	it('should have correct class when isSuccess passed from Form Control', () => {
		render(
			<FormControl id="control" isSuccess>
				<TextInput />
			</FormControl>
		);

		const input = screen.getByRole('textbox');

		expect(input).toHaveClass('is-success');
	});
});
