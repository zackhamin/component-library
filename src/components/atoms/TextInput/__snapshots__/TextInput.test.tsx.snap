// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders correctly 1`] = `
.c1 {
  height: 4.8rem;
  width: 100%;
  padding: 1.6rem;
  font-size: 1.6rem;
  font-family: Helvetica,Arial,sans-serif;
  color: #333333;
  border: 1px solid #333333;
  border-radius: 0.4rem;
}

.c1::-webkit-input-placeholder {
  color: #979797;
}

.c1::-moz-placeholder {
  color: #979797;
}

.c1:-ms-input-placeholder {
  color: #979797;
}

.c1::placeholder {
  color: #979797;
}

.c1:disabled {
  background-color: #EEEEEF;
  border-color: #979797;
}

.c1:focus-visible {
  outline: none;
  border-color: #00594C;
  box-shadow: 0px 0px 6px rgba(0,89,76,0.4);
}

.c1.is-invalid {
  border-color: #D61616;
  padding-right: 5rem;
}

.c1.is-success {
  padding-right: 5rem;
}

.c0 {
  position: relative;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1 additional-class"
      id="inputID"
      name="inputName"
      placeholder="Placeholder text"
      type="text"
      value=""
    />
  </div>
</div>
`;
